### PR TITLE
Add replaceFirst function

### PR DESCRIPTION
**Pull request by yuandagits**
Summary:
ReplaceFirst function will replace the first found instance of the search with the replacement. If the search is empty, replaceFirst will prepend the replacement on the input string.

We simply leverage the existing replace function and pass in a replaceFirst bool flag.
Differential Revision: D64475513

### CODE REVIEW

**Review by bikramSingh91 — COMMENTED** [review]
@yuandagits A lot of the implementation is the same as `replace` UDF,  to avoid code duplication would it be possible to refactor the `replace` udf's implementation to also work for a single replacement?

**Comment by yuandagits** [comment]
> @yuandagits A lot of the implementation is the same as `replace` UDF, to avoid code duplication would it be possible to refactor the `replace` udf's implementation to also work for a single replacement?

Done

File: velox/functions/prestosql/StringFunctions.cpp
Comment (bikramSingh91) [inline velox/functions/prestosql/StringFunctions.cpp:444]: Can you also de-duplicate the code for ReplaceFirst and reuse the one for replace?

You can do this by changing the constructor of replace to take a bool like `Replace(bool isReplaceFirst)` then using this wherever the implementation diverges. Also, add the same bool in `signatures(bool isReplaceFirst)` and only register the signature relevant to it. Finally, while registering you can pass in true and false corresponding to ReplaceFirst and Replace respectively. For a code reference on how the same vector function is used for different UDFs, you can checkout ArraySort.cpp

**Review by bikramSingh91 — CHANGES_REQUESTED** [review]

File: velox/functions/prestosql/StringFunctions.cpp
Comment (yuandagits) [reply velox/functions/prestosql/StringFunctions.cpp:444]: I can add it. I think sigantures would be a no-op, right? replace anyways needs 
```
        exec::FunctionSignatureBuilder()
            .returnType("varchar")
            .argumentType("varchar")
            .argumentType("varchar")
            .argumentType("varchar")
            .build(),
    ```
    which is what replaceFirst needs

Comment (yuandagits) [reply velox/functions/prestosql/StringFunctions.cpp:444]: nvm I misunderstood, i got it, we need to apply the delta during registration. thank you!

File: velox/functions/lib/string/tests/StringImplTest.cpp
Comment (bikramSingh91) [inline velox/functions/lib/string/tests/StringImplTest.cpp:557]: can you add a test using unicode?

File: velox/functions/prestosql/tests/StringFunctionsTest.cpp
Comment (bikramSingh91) [inline velox/functions/prestosql/tests/StringFunctionsTest.cpp:1565]: Just noticed this, you should be able to re-use `StringFunctionsTest::testReplaceInPlace` by adding another input param `functionName` then using it instead of the hardcoded `replace` in its implementation. The test logic is fairly involved here, so would be good to keep it shared.

**Review by bikramSingh91 — APPROVED** [review]
Looks good, just 2 more nits in the test. Accepting to unblock, but please address them before merging.

**Pull request closed without merging**
